File: problems/C++/26_smallest_range_i.cpp
// Problem Statement

// Given an array A of integers, for each integer A[i] we may choose any x with -K <= x <= K, and add x to A[i].
//
// After this process, we have some array B.
//
// Return the smallest possible difference between the maximum value of B and the minimum value of B.

#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    int smallestRangeI(vector<int>& A, int K) {
          int minVal = INT_MAX,maxVal = INT_MIN;
          if(A.size()<2)
            return 0;

          for(auto val : A){
              minVal = min(minVal,val);
              maxVal = max(maxVal,val);
          }

          if(maxVal-K<=minVal+K)
            return 0;

          return (maxVal-K)-(minVal+K);
    }
};

int main(){
  vector<int> temp = {2,7,2};
  Solution sol;
  cout<<sol.smallestRangeI(temp,1)<<endl;
  return 0;
}

File: problems/Kotlin/26_smallest_range_i.kt
// Problem Statement

// Given an array A of integers, for each integer A[i] we may choose any x with -K <= x <= K, and add x to A[i].
//
// After this process, we have some array B.
//
// Return the smallest possible difference between the maximum value of B and the minimum value of B.

class Solution {
    fun smallestRangeI(A: IntArray, K: Int): Int {
        var maxVal:Int = Int.MIN_VALUE;
        var minVal:Int = Int.MAX_VALUE;

        for(value in A){
            maxVal = maxOf(maxVal,value);
            minVal = minOf(minVal,value);
        }

        if(((maxVal-K)-(minVal+K))<=0){
            return 0;
        }else{
            return (maxVal-K)-(minVal+K);
        }
    }
}

fun main(args:Array<String>){
  val vec:IntArray = intArrayOf(1,3,6);
  val sol:Solution = Solution();

  println(sol.smallestRangeI(vec,3));
}
